File: java/AlgorithmAndDataStructure/src/main/java/algorithms/ModifyingSequence.java
package algorithms;

import functionObjects.Generator;
import functionObjects.UnaryOp;
import functionObjects.UnaryPredicate;

import java.util.*;

/**
 * Created by lzx on 15-2-16.
 */
public class ModifyingSequence {
	public static void copy(Collection a, Collection b) {
		b.addAll(a);
	}

	public static void copy_if(Collection a, Collection b, UnaryPredicate f) {
		for (Object o : a) {
			if (f.call(o))
				b.add(o);
		}
	}

	public static <T> void copy_n(Collection<T> a, int count, Collection b) {
		for (T o : a) {
			if (count == 0)
				break;
			b.add(o);
			count--;
		}
	}

	public static void copy_backward(List a, List b) {
		int idxB = b.size() - 1;
		ListIterator iterA = a.listIterator(a.size());
		while (iterA.hasPrevious() && idxB >= 0) {
			b.set(idxB, iterA.previous());
			idxB--;
		}
	}

	public static void move(Collection a, Collection b) {
		copy(a, b);
	}

	public static void move_backward(List a, List b) {
		copy_backward(a, b);
	}

	public static <T> void fill(List<T> a, T value) {
		int len = a.size();
		for (int i = 0; i < len; i++) {
			a.set(i, value);
		}
	}

	public static <T> void fill_n(List<T> a, int count, T value) {
		for (int i = 0; i < count; i++) {
			a.set(i, value);
		}
	}

	public static void transform(Collection a, Collection b, UnaryOp f) {
		for (Object o : a) {
			b.add(f.call(o));
		}
	}

	public static void generate(List a, Generator g) {
		int len = a.size();
		for (int i = 0; i < len; i++) {
			a.set(i, g.call());
		}
	}

	public static void generate_n(List a, int count, Generator g) {
		for (int i = 0; i < count; i++) {
			a.set(i, g.call());
		}
	}

	public static boolean remove(Collection a, Object match) {
		boolean removed = false;
		Iterator i = a.iterator();
		while (i.hasNext()) {
			if (i.next().equals(match)) {
				i.remove();
				removed = true;
			}
		}
		return removed;
	}

	public static boolean removeIf(Collection a, UnaryPredicate f) {
		boolean removed = false;
		Iterator i = a.iterator();
		while (i.hasNext()) {
			if (f.call(i.next())) {
				i.remove();
				removed = true;
			}
		}
		return removed;
	}

	public static <T> void removeCopy(Collection<T> a, Collection<T> b, T
			match) {
		Iterator<T> i = a.iterator();
		while (i.hasNext()) {
			T tmp = i.next();
			if (tmp.equals(match)) {
				i.remove();
				b.add(tmp);
			}
		}
	}

	public static <T> void removeCopyIf(Collection<T> a, Collection<T> b,
	                                 UnaryPredicate f) {
		Iterator<T> i = a.iterator();
		while (i.hasNext()) {
			T tmp = i.next();
			if (f.call(tmp)) {
				i.remove();
				b.add(tmp);
			}
		}
	}

	public static <T> boolean replace(List<T> a,  T before, T after) {
		boolean replaced = false;
		int len = a.size();
		for (int i = 0; i < len; i++) {
			T tmp = a.get(i);
			if (before == null ? tmp == null : tmp.equals(before)) {
				a.set(i, after);
				replaced = true;
			}
		}
		return replaced;
	}

	public static <T> boolean replaceIf(List<T> a, UnaryPredicate f, T after) {
		boolean replaced = false;
		int len = a.size();
		for (int i = 0; i < len; i++) {
			if (f.call(a.get(i))) {
				a.set(i, after);
				replaced = true;
			}
		}
		return replaced;
	}

	public static <T> boolean replaceCopy(List<T> a,  List<T> b, T before, T after) {
		boolean replaced = false;
		for (T tmp : a) {
			if (before == null ? tmp == null : tmp.equals(before)) {
				b.add(after);
				replaced = true;
			}
			else {
				b.add(tmp);
			}
		}
		return replaced;
	}

	public static <T> boolean replaceCopyIf(List<T> a, List<T> b,
	                                        UnaryPredicate f, T after) {
		boolean replaced = false;
		for (T tmp : a) {
			if (f.call(tmp)) {
				b.add(after);
				replaced = true;
			}
			else {
				b.add(tmp);
			}
		}
		return replaced;
	}

	// it is impossible to implement a generic swap function in Java
	public static <T> void range_swap(List<T> a, List<T> b) {
		int len = b.size() > a.size() ? a.size() : b.size();
		for (int i = 0; i < len; i++) {
			T tmp = a.get(i);
			a.set(i, b.get(i));
			b.set(i, tmp);
		}
	}

	public static <T> void reverse(List<T> a) {
		int size = a.size() - 1;
		int middle = (size + 1) / 2;
		for (int i = 0; i < middle; i++) {
			T tmp = a.get(i);
			a.set(i, a.get(size - i));
			a.set(size - i, tmp);
		}
	}

	public static void reverseCopy(List a, List b) {
		ListIterator i = a.listIterator(a.size());
		while (i.hasPrevious()) {
			b.add(i.previous());
		}
	}

	// Two version of rotate, one for List supports RandomAccess, one for
	// List not supports RandomAccess. See doc of java.util.Collections#rotate.
	// I only implement the second one, since it is easier than the first
	// one :)
	public static <T> void rotate(List<T> a, int distance) {
		int size = a.size();
		if (size == 0)
			return;
		int pivot = -distance % size;
		if (pivot < 0)
			pivot += size;
		if (pivot == 0)
			return;
		reverse(a.subList(0, pivot));
		reverse(a.subList(pivot, size));
		reverse(a);
	}

	public static <T> void rotateCopy(List<T> a, List<T> b, int distance) {
		b.addAll(a);
		rotate(b, distance);
	}

	public static <T> void shuffle(List<T> a) {
		int size = a.size();
		Random rd = new Random();
		for (int i = 0; i < size; i++) {
			int randomNum = rd.nextInt(size);
			T tmp = a.get(i);
			a.set(i, a.get(randomNum));
			a.set(randomNum, tmp);
		}
	}

	public static <T> void unique(Collection<T> a) {
		Iterator<T> i = a.iterator();
		if (!i.hasNext()) {
			return;
		}
		T head = i.next();
		while (i.hasNext()) {
			T cur = i.next();
			if (head.equals(cur)) {
				i.remove();
			}
			else {
				head = cur;
			}
		}
	}

	public static <T> void uniqueCopy(Collection<T> a, Collection<T> b) {
		Iterator<T> i = a.iterator();
		if (!i.hasNext()) {
			return;
		}
		T head = i.next();
		b.add(head);
		while (i.hasNext()) {
			T cur = i.next();
			if (!head.equals(cur)) {
				b.add(cur);
				head = cur;
			}
		}
	}
}

File: scala/src/main/scala/algorithm/ModifyingSequence.scala
package algorithm

import collection.mutable.ArrayBuffer
import collection.mutable.Iterable
import collection.mutable.Seq
import util.Random

/**
 * Created by lzx on 15-4-18.
 * Although this object is named with Modifying, but scala recommends people to 
 * code in functional style. So I try to implement majority of them in non-modifying way.
 */
object ModifyingSequence {
  def Copy[T](a: Iterable[T], b: Iterable[T]) = {
    (for ((x, y) <- a.zip(b)) yield y) ++ a.slice(b.size, a.size)
  }

  def CopyIf[T](a: Iterable[T], b: Iterable[T], f: T => Boolean) = {
    this.Copy(a, b.filter(f))
  }

  def CopyN[T](a: Iterable[T], b: Iterable[T], count: Int) = {
    this.Copy(a, b.take(count))
  }

  def CopyBackward[T](a: Seq[T], b: Seq[T]) = {
    val aSize = a.size
    val bSize = b.size
    if (aSize <= bSize) {
      b.slice(bSize - aSize, bSize)
    }
    else {
      a.slice(0, aSize - bSize) ++ b
    }
  }
  // Move is the same as Copy, and MoveBackward is the same as CopyBackward.

  def Fill[T](c: Iterable[T], e: T) = {
    c.map((x: T) => e)
  }

  def FillN[T](c: Seq[T], count: Int, e: T) = {
    val cSize = c.size
    if (cSize <= count) {
      this.Fill(c, e)
    }
    else {
      Seq.fill(count)(e) ++ c.slice(count, cSize)
    }
  }

  def Transform[T](c: Iterable[T], f: T => T) = {
    c.map(f)
  }

  def Generate[T](c: Seq[T], g: () => T) {
    for (i <- 0 until c.size)
      c(i) = g()
  }

  // instead of changing the value of the first N elements of a Sequence,
  // return a new Sequence with size of N
  def GenerateN[T](N: Int, g: () => T) = {
    for (i <- 0 until N) yield g()
  }

  def Remove[T](c: Iterable[T], value: T) = {
    c.filter((x: T) => x != value)
  }

  def RemoveIf[T](c: Iterable[T], f: T => Boolean) = {
    c.filterNot(f)
  }

  // return two Iterable, one is elements removed, the other is elements remained
  def RemoveCopyIf[T](c: Iterable[T], f: T => Boolean) = {
    c.partition(f)
  }

  def RemoveCopy[T](c: Iterable[T], value: T) = {
    this.RemoveCopyIf(c, (x: T) => x == value)
  }

  def Replace[T](c: Seq[T], before: T, after: T) {
    for (i <- 0 until c.size) {
      if (c(i) == before) {
        c(i) = after
      }
    }
  }

  def ReplaceIf[T](c: Seq[T], f: T => Boolean, after: T) {
    for (i <- 0 until c.size) {
      if (f(c(i))) {
        c(i) = after
      }
    }
  }

  def ReplaceCopy[T](c: Seq[T], before: T, after: T) = {
    c.map((x: T) => if (x == before) after else x)
  }

  def ReplaceCopyIf[T](c: Seq[T], f: T => Boolean, after: T) = {
    c.map((x: T) => if (f(x)) after else x)
  }

  def RangeSwap[T](a: Seq[T], b: Seq[T]) {
    var i = 0
    for ((x, y) <- a.zip(b)) {
      a(i) = y
      b(i) = x
      i += 1
    }
  }

  def Reverse[T](c: Seq[T]) {
    val size = c.size - 1
    val middle = (size + 1) / 2
    for (i <- 0 until middle) {
      val tmp = c(i)
      c(i) = c(size - i)
      c(size - i) = tmp
    }
  }

  def ReverseCopy[T](c: Seq[T]) = {
    c.reverseMap((x: T) => x)
  }

  def Rotate[T](c: Seq[T], dist: Int) {
    val cSize = c.size
    if (cSize != 0 && dist != 0) {
      var idx = -dist % cSize
      if (idx < 0)
        idx += cSize
      val pivot1 = idx / 2
      for (i <- 0 until pivot1) {
        val tmp = c(i)
        c(i) = c(idx - i - 1)
        c(idx - i - 1) = tmp
      }
      val pivot2 = (cSize + idx) / 2
      for (i <- idx until pivot2) {
        val tmp = c(i)
        c(i) = c(cSize - 1 - i + idx)
        c(cSize - 1 - i + idx) = tmp
      }
      this.Reverse(c)
    }
  }

  def RotateCopy[T](c: Seq[T], dist: Int) = {
    val cSize = c.size
    var idx = (cSize - dist) % cSize
    if (idx < 0)
      idx = idx + cSize
    c.slice(idx, cSize) ++ c.slice(0, idx)
  }

  def Shuffle[T](c: Seq[T]) = {
    val rd = new Random()
    val size = c.size
    for (i <- 0 until size) {
      val randNum = rd.nextInt(size)
      val tmp = c(i)
      c(i) = c(randNum)
      c(randNum) = tmp
    }
    c
  }

  // works like distince
  def UniqueCopy[T](c: Seq[T]) =  {
    if (c.size <= 1) {
      c
    }
    else {
      var head = c.head
      var start = head 
      val f = (x: T) => {
        if (x != head) {
          head = x
          true
        }
        else {
          false
        }
      }
      start +: c.filter(f)
    }
  }

  def Unique[T](c: ArrayBuffer[T]) {
    val uniq = this.UniqueCopy(c)
    c.clear
    c.appendAll(uniq)
  }
}
